--- Day 11: Plutonian Pebbles ---
The ancient civilization on Pluto was known for its ability to manipulate spacetime, and while The Historians explore their infinite corridors, you've noticed a strange set of physics-defying stones.

At first glance, they seem like normal stones: they're arranged in a perfectly straight line, and each stone has a number engraved on it.

The strange part is that every time you blink, the stones change.

Sometimes, the number engraved on a stone changes. Other times, a stone might split in two, causing all the other stones to shift over a bit to make room in their perfectly straight line.

As you observe them for a while, you find that the stones have a consistent behavior. Every time you blink, the stones each simultaneously change according to the first applicable rule in this list:

If the stone is engraved with the number 0, it is replaced by a stone engraved with the number 1.
If the stone is engraved with a number that has an even number of digits, it is replaced by two stones. The left half of the digits are engraved on the new left stone, and the right half of the digits are engraved on the new right stone. (The new numbers don't keep extra leading zeroes: 1000 would become stones 10 and 0.)
If none of the other rules apply, the stone is replaced by a new stone; the old stone's number multiplied by 2024 is engraved on the new stone.
No matter how the stones change, their order is preserved, and they stay on their perfectly straight line.

How will the stones evolve if you keep blinking at them? You take a note of the number engraved on each stone in the line (your puzzle input).

If you have an arrangement of five stones engraved with the numbers 0 1 10 99 999 and you blink once, the stones transform as follows:

The first stone, 0, becomes a stone marked 1.
The second stone, 1, is multiplied by 2024 to become 2024.
The third stone, 10, is split into a stone marked 1 followed by a stone marked 0.
The fourth stone, 99, is split into two stones marked 9.
The fifth stone, 999, is replaced by a stone marked 2021976.
So, after blinking once, your five stones would become an arrangement of seven stones engraved with the numbers 1 2024 1 0 9 9 2021976.

Here is a longer example:

Initial arrangement:
125 17

After 1 blink:
253000 1 7

After 2 blinks:
253 0 2024 14168

After 3 blinks:
512072 1 20 24 28676032

After 4 blinks:
512 72 2024 2 0 2 4 2867 6032

After 5 blinks:
1036288 7 2 20 24 4048 1 4048 8096 28 67 60 32

After 6 blinks:
2097446912 14168 4048 2 0 2 4 40 48 2024 40 48 80 96 2 8 6 7 6 0 3 2
In this example, after blinking six times, you would have 22 stones. After blinking 25 times, you would have 55312 stones!

Consider the arrangement of stones in front of you. How many stones will you have after blinking 25 times?

To begin, get your puzzle input.

Answer: 
----------------------------------
3 rules in order:
1. if the stone is 0, turn into 1
2. if the stone's digit number is even, split the digits evenly into two
3. otherwise, time the number by 2024

for each stone, extract it in reverse order, so the splitting stones won't 

In [1]:
##*****************************
# Part I
##*****************************
strStones = "6563348 67 395 0 6 4425 89567 739318"
lstStones = list(map(int, strStones.split(' ')))    ## convert the elements to int type

def blink(lstStones) :
    totalStones = len(lstStones)
    for i in range(totalStones) :
        idx = totalStones-i-1           ## loop the stone backwards
        stone = lstStones[idx]          ## record the stone for this loop
        #print(stone)
        if stone == 0 :                 ## if stone is 0, turn into 1
            lstStones[idx] = 1
        elif len(str(stone))%2 == 0 :   ## if stone's digit number is even, split the digits evenly into two
            lstStones[idx] = int(str(stone)[:int(len(str(stone))/2)])
            lstStones.insert(idx+1, int(str(stone)[int(len(str(stone))/2):]))
        else :                          ## otherwise, time the number by 2024
            lstStones[idx] *= 2024
        #print(lstStones)
    #print(lstStones)

##=================================
## Part I Main Program Start Here
##=================================
x = 25
## blink x times
for i in range(x) :
    blink(lstStones)

##
print(f"Total Stones after {x} blinks : {len(lstStones)}")

Total Stones after 25 blinks : 184927


--- Part Two ---
The Historians sure are taking a long time. To be fair, the infinite corridors are very large.

How many stones would you have after blinking a total of 75 times?

Answer: 
-----------------------------------------------
need to consider the performance and memory consumation
1. Since each stone's next transform is independent, we can process them in different order, i.e. in sorted order
2. group the same stone numbers and create a different list to store the occurence of the number
3. only process/blink a number once in the same lineup
4. reshuffle the unique number list and their occurence list
5. from the analysis below, after 25 blinks, the unique number is around 454, instead of processing 184,927 stones

In [ ]:
##*****************************
# Part II - Testing - analsis
##*****************************
strStones = "6563348 67 395 0 6 4425 89567 739318" #"1"
lstStones = list(map(int, strStones.split(' ')))    ## convert the elements to int type

def blink(lstStones) :
    totalStones = len(lstStones)
    for i in range(totalStones) :
        idx = totalStones-i-1           ## loop the stone backwards
        stone = lstStones[idx]          ## record the stone for this loop
        #print(stone)
        if stone == 0 :                 ## if stone is 0, turn into 1
            lstStones[idx] = 1
        elif len(str(stone))%2 == 0 :   ## if stone's digit number is even, split the digits evenly into two
            lstStones[idx] = int(str(stone)[:int(len(str(stone))/2)])
            lstStones.insert(idx+1, int(str(stone)[int(len(str(stone))/2):]))
        else :                          ## otherwise, time the number by 2024
            lstStones[idx] *= 2024
        #print(lstStones)
    #print(lstStones)
    #print(len(lstStones))

##=======================================
## compare of actual size to the dedups
##=======================================
x = 25
## blink x times
for i in range(x) :
    blink(lstStones)
    #print(lstStones)
    orgSize = len(lstStones)
    dedupSize = len(list(set(lstStones)))
    print(f"orgSize={orgSize}, dedupSize={dedupSize}, Reduced={orgSize-dedupSize}")
    #print(f"Total Stones after {i+1} blinks : {len(lstStones)}")

##
#print(f"Total Stones after {x} blinks : {len(lstStones)}")

orgSize=11, dedupSize=11, Reduced=0
orgSize=14, dedupSize=13, Reduced=1
orgSize=19, dedupSize=18, Reduced=1
orgSize=31, dedupSize=28, Reduced=3
orgSize=50, dedupSize=37, Reduced=13
orgSize=69, dedupSize=45, Reduced=24
orgSize=96, dedupSize=55, Reduced=41
orgSize=146, dedupSize=72, Reduced=74
orgSize=238, dedupSize=94, Reduced=144
orgSize=380, dedupSize=92, Reduced=288
orgSize=548, dedupSize=102, Reduced=446
orgSize=786, dedupSize=112, Reduced=674
orgSize=1243, dedupSize=127, Reduced=1116
orgSize=1871, dedupSize=151, Reduced=1720
orgSize=2875, dedupSize=173, Reduced=2702
orgSize=4295, dedupSize=200, Reduced=4095
orgSize=6456, dedupSize=209, Reduced=6247
orgSize=10000, dedupSize=219, Reduced=9781
orgSize=14984, dedupSize=232, Reduced=14752
orgSize=23015, dedupSize=257, Reduced=22758
orgSize=34805, dedupSize=290, Reduced=34515
orgSize=52300, dedupSize=341, Reduced=51959
orgSize=80345, dedupSize=359, Reduced=79986
orgSize=121467, dedupSize=386, Reduced=121081
orgSize=184927, dedupSize=454,

In [13]:
a=2024
cnt=0
prev = -1
for i in range(20): # 1,000,000,000
    #x = 0
    #print(f"i={i}, i*2024={i*a}, len()={len(str(a*i))}")
    seed = int(10**i/2024)+1
    print(f"seed={seed-1}, seed*2024={(seed-1)*2024}, seedLen={len(str(seed-1))}, 2024Len={len(str((seed-1)*2024))}")
    print('-----------------')
    print(f"seed={seed}, seed*2024={seed*2024}, seedLen={len(str(seed))}, 2024Len={len(str(seed*2024))}")


    # if len(str(a*i))%2==1:
    #     print(i)
    #     cnt += 1

print(f"count={cnt}")

seed=0, seed*2024=0, seedLen=1, 2024Len=1
-----------------
seed=1, seed*2024=2024, seedLen=1, 2024Len=4
seed=0, seed*2024=0, seedLen=1, 2024Len=1
-----------------
seed=1, seed*2024=2024, seedLen=1, 2024Len=4
seed=0, seed*2024=0, seedLen=1, 2024Len=1
-----------------
seed=1, seed*2024=2024, seedLen=1, 2024Len=4
seed=0, seed*2024=0, seedLen=1, 2024Len=1
-----------------
seed=1, seed*2024=2024, seedLen=1, 2024Len=4
seed=4, seed*2024=8096, seedLen=1, 2024Len=4
-----------------
seed=5, seed*2024=10120, seedLen=1, 2024Len=5
seed=49, seed*2024=99176, seedLen=2, 2024Len=5
-----------------
seed=50, seed*2024=101200, seedLen=2, 2024Len=6
seed=494, seed*2024=999856, seedLen=3, 2024Len=6
-----------------
seed=495, seed*2024=1001880, seedLen=3, 2024Len=7
seed=4940, seed*2024=9998560, seedLen=4, 2024Len=7
-----------------
seed=4941, seed*2024=10000584, seedLen=4, 2024Len=8
seed=49407, seed*2024=99999768, seedLen=5, 2024Len=8
-----------------
seed=49408, seed*2024=100001792, seedLen=5, 2024L

In [20]:
from collections import Counter

##*****************************
# Part II - performance tuning
##*****************************
strStones = "6563348 67 395 0 6 4425 89567 739318"
lstStones = list(map(int, strStones.split(' ')))    ## convert the elements to int type
dicStones = dict(Counter(sorted(lstStones)))        ## convert into dic with counts in value (sorted keys too)
#print(dicStones)

def blink(dicInput) :
    #totalStones = len(dicInput)
    dicOutput = {}
    for num, cnt in dicInput.items() :
        #idx = totalStones-i-1           ## loop the stone backwards
        stone = int(num)                ## record the stone for this loop
        #print(stone)
        if stone == 0 :                 ## if stone is 0, turn into 1
            dicOutput[1] = cnt if 1 not in dicOutput else cnt+dicOutput[1]
        elif len(str(stone))%2 == 0 :   ## if stone's digit number is even, split the digits evenly into two
            leftSplit = int(str(stone)[:int(len(str(stone))/2)])
            rightSplit = int(str(stone)[int(len(str(stone))/2):])
            dicOutput[leftSplit] = cnt if leftSplit not in dicOutput else cnt+dicOutput[leftSplit]
            dicOutput[rightSplit] = cnt if rightSplit not in dicOutput else cnt+dicOutput[rightSplit]
        else :                          ## otherwise, time the number by 2024
            dicOutput[stone*2024] = cnt if stone*2024 not in dicOutput else cnt+dicOutput[stone*2024]
        #print(dicOutput)
    #print(dicInput)

    return dict(sorted(dicOutput.items()))

##=================================
## Part II Main Program Start Here
##=================================
x = 75      ## blink x times

print(dicStones)
for i in range(x):  ## repeat x times
    dicStones= blink(dicStones)
    print(f"{i} iteration with {len(dicStones)} distinct stone numbers")

# add up all values in the dictionary
print(f"Number of Stones={sum(dicStones.values())}")



{0: 1, 6: 1, 67: 1, 395: 1, 4425: 1, 89567: 1, 739318: 1, 6563348: 1}
0 iteration with 11 distinct stone numbers
1 iteration with 13 distinct stone numbers
2 iteration with 18 distinct stone numbers
3 iteration with 28 distinct stone numbers
4 iteration with 37 distinct stone numbers
5 iteration with 45 distinct stone numbers
6 iteration with 55 distinct stone numbers
7 iteration with 72 distinct stone numbers
8 iteration with 94 distinct stone numbers
9 iteration with 92 distinct stone numbers
10 iteration with 102 distinct stone numbers
11 iteration with 112 distinct stone numbers
12 iteration with 127 distinct stone numbers
13 iteration with 151 distinct stone numbers
14 iteration with 173 distinct stone numbers
15 iteration with 200 distinct stone numbers
16 iteration with 209 distinct stone numbers
17 iteration with 219 distinct stone numbers
18 iteration with 232 distinct stone numbers
19 iteration with 257 distinct stone numbers
20 iteration with 290 distinct stone numbers
21 it

In [14]:
stone=6
cnt=1
dicOutput={1:1}
newstone = stone *2024
if newstone not in dicOutput:
    dicOutput[newstone] = cnt
else :
    dicOutput[newstone] += cnt
print(dicOutput)

{1: 1, 12144: 1}
